src/mame/misc/babysuprem.cpp

```diff
@@ -33,45 +46,364 @@ class bsuprem_state : public driver_device
+
+	uint8_t dsw2_u13_switches_r(offs_t offset);
+	void u9_w(offs_t offset, uint8_t data);
+	void u10_w(offs_t offset, uint8_t data);
+
+	virtual void machine_reset() override;
+	virtual void machine_start() override;
+
+	INTERRUPT_GEN_MEMBER(mains_irq);
+
 };
 
 static INPUT_PORTS_START( bsuprem )
-	PORT_START("DSW1")
-	PORT_DIPUNKNOWN_DIPLOC(0x01, 0x01, "SW1:1")
-	PORT_DIPUNKNOWN_DIPLOC(0x02, 0x02, "SW1:2")
-	PORT_DIPUNKNOWN_DIPLOC(0x04, 0x04, "SW1:3")
-	PORT_DIPUNKNOWN_DIPLOC(0x08, 0x08, "SW1:4")
+	PORT_START("DSW1_SWITCHES2")
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_BUTTON3) PORT_NAME("Right stop")
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_BUTTON2) PORT_NAME("Middle stop")
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_BUTTON1) PORT_NAME("Left stop")
+	PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_BUTTON4)
+	PORT_DIPNAME(0x80, 0x80, "SW1-3 (Test 4)")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x80, DEF_STR( Off ))
+	PORT_DIPNAME(0x40, 0x40, "SW1-1 (Test 2)")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x40, DEF_STR( Off ))
+	PORT_DIPNAME(0x20, 0x20, "SW1-2 (Test 3)")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x20, DEF_STR( Off ))
+	PORT_DIPNAME(0x10, 0x10, "SW1-4 (Test 5)")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x10, DEF_STR( Off ))
 
 	PORT_START("DSW2")
-	PORT_DIPUNKNOWN_DIPLOC(0x01, 0x01, "SW2:1")
-	PORT_DIPUNKNOWN_DIPLOC(0x02, 0x02, "SW2:2")
-	PORT_DIPUNKNOWN_DIPLOC(0x04, 0x04, "SW2:3")
-	PORT_DIPUNKNOWN_DIPLOC(0x08, 0x08, "SW2:4")
-	PORT_DIPUNKNOWN_DIPLOC(0x10, 0x10, "SW2:5")
-	PORT_DIPUNKNOWN_DIPLOC(0x20, 0x20, "SW2:6")
-	PORT_DIPUNKNOWN_DIPLOC(0x40, 0x40, "SW2:7")
-	PORT_DIPUNKNOWN_DIPLOC(0x80, 0x80, "SW2:8")
+	PORT_DIPNAME(0x80, 0x80, "SW2-2 (Door)")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x80, DEF_STR( Off ))
+	PORT_DIPNAME(0x40, 0x40, "SW2-4")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x40, DEF_STR( Off ))
+	PORT_DIPNAME(0x20, 0x20, "SW2-6 Auto start")
+	PORT_DIPSETTING(   0x00, DEF_STR( Yes ))
+	PORT_DIPSETTING(   0x20, DEF_STR( No ))
+	PORT_DIPNAME(0x10, 0x10, "SW2-8 Percent")
+	PORT_DIPSETTING(   0x00, "80%")
+	PORT_DIPSETTING(   0x10, "85%")
+	PORT_DIPNAME(0x08, 0x08, "SW2-7")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x08, DEF_STR( Off ))
+	PORT_DIPNAME(0x04, 0x04, "SW2-5 Test")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x04, DEF_STR( Off ))
+	PORT_DIPNAME(0x02, 0x02, "SW2-3 Instant start")
+	PORT_DIPSETTING(   0x00, DEF_STR( Yes ))
+	PORT_DIPSETTING(   0x02, DEF_STR( No ))
+	PORT_DIPNAME(0x01, 0x01, "SW2-1 (Bet/Collect)")
+	PORT_DIPSETTING(   0x00, DEF_STR( On ))
+	PORT_DIPSETTING(   0x01, DEF_STR( Off ))
+
+	PORT_START("U13_SWITCHES")
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_BUTTON5)  PORT_NAME("Bet/Collect")
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_BUTTON6)  PORT_NAME("Start")
+	PORT_BIT(0x80, IP_ACTIVE_LOW, IPT_SERVICE)  PORT_NAME("Door") PORT_CODE(KEYCODE_D) PORT_TOGGLE
```

Comment:
Same here, please use appropriate input types, and use proper markup for DIP switches.